data_structures/linked_list/circular_doubly_linked_list.c

```diff
@@ -0,0 +1,307 @@
+        new_node->prev = temp;
+        temp->next = new_node;
+        head = new_node;
+        return head;
+    }
+}
+
+/**
+ * @brief Insert the node at end of list
+ *
+ * @param head      starting pointer of list
+ * @param data      The data which we want to insert into list
+ * @return ListNode*
+ */
+ListNode *insert_at_last(ListNode *head, uint64_t data)
+{
+    if (head == NULL)
+    {
+        head = create_list(data);
+        return head;
+    }
+    else
+    {
+        ListNode *temp1, *temp2;
+        ListNode *new_node = create_list(data);
+        temp1 = head;
+        temp2 = head->prev;
+        new_node->prev = temp2;
+        new_node->next = temp1;
+        temp1->prev = new_node;
+        temp2->next = new_node;
+        return head;
+    }
+}
+
+/**
+ * @brief  Function for deletion of the first node in list
+ *
+ * @param head      starting pointer of list
+ * @return ListNode*
+ */
+ListNode *delete_first_node(ListNode *head)
+{
+    if (head == NULL)
+    {
+        printf("list is Empty\n");
+        return head;
+    }
+    ListNode *temp1, *temp2;
+    temp1 = head;
+    temp2 = temp1->prev;
+    if (temp1 == temp2)
+    {
+        free(temp2);
+        head = NULL;
+        printf("\nNode succesfully deleted\n\n");
+        return head;
+    }
+    temp2->next = temp1->next;
+    (temp1->next)->prev = temp2;
+    head = temp1->next;
+    free(temp1);
+    temp1 = NULL;
+    printf("\nNode succesfully deleted\n\n");
+    return head;
+}
+
+/**
+ * @brief Function for deletion of the last node in list
+ *
+ * @param head      starting pointer of list
+ * @return ListNode*
+ */
+ListNode *delete_last_node(ListNode *head)
+{
+    if (head == NULL)
+    {
+        printf("list is Empty\n");
+        return head;
+    }
+
+    ListNode *temp1, *temp2;
+    temp1 = head;
+    temp2 = temp1->prev;
+    if (temp1 == temp2)
+    {
+        free(temp2);
+        head = NULL;
+        printf("\nNode succesfully deleted\n\n");
+        return head;
+    }
+    (temp2->prev)->next = temp1;
+    temp1->prev = temp2->prev;
+    free(temp2);
+    temp2 = NULL;
+    printf("\nNode succesfully deleted\n\n");
+    return head;
+}
+
+/**
+ * @brief The function to reverse the linked list
+ *
+ * @param head      starting pointer of list
+ * @return ListNode*
+ */
+ListNode *reverseLL(ListNode *head)
+{
+    if (!head)
+        return NULL;
+
+    ListNode *new_head = NULL;
+    ListNode *last = head->prev;
+    ListNode *curr = last, *prev;
+
+    // traverse list in backward direction
+    while (curr->prev != last)
+    {
+        prev = curr->prev;
+        new_head = insert_at_last(new_head, curr->value);
+        curr = prev;
+    }
+    new_head = insert_at_last(new_head, curr->value);
+
+    return new_head;
+}
+
+/**
+ * @brief The function that will return current size of list
+ *
+ * @param head      starting pointer of list
+ * @return int      size of list
+ */
+int getsize(ListNode *head)
```

Comment:
What would happen if an empty list is inputted into this function?

Reply:
I'll add
```cpp
if(!head) return 0;
```